### PR TITLE
Implement several abilities and 17 cards

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Implement abilities
    - Reborn: Reborn is an ability that causes a minion to be resummoned with 1 Health the first time it's destroyed.
    - Quest: A Quest is a spell ability and type of spell card which grants a conditional reward. When played, Quests are displayed above the controlling hero's portrait, similar to Secrets. When the condition of the Quest is met, the Quest's reward is given to the player. Unlike Secrets, Quests can be seen by the opponent when they are hovered over, the Quest's progress is displayed to both players, and their conditions are met by the possessing player rather than by the opponent.
    - Sidequest: Sidequests are a spell ability and type of spell card which grant a conditional reward. When played, Sidequests are displayed above the controlling hero's portrait, similar to Secrets and Quests. When the condition of the Sidequest is met, the Sidequest's reward is given to the player. Sidequests can be seen by the opponent when they are hovered over, the Sidequest's progress is displayed to both players. Unlike Quests, Sidequests come in rarities other than Legendary, do not start in the opening mulligan, and multiple Sidequests can be active at a time.
    - Discover: Discover is an ability that allows the player to choose one of three different cards. The three cards are randomly selected from a pool of all eligible cards, as stated by the card text. If not stated, the pool of cards consists of all eligible neutral and class cards matching the player's current hero class, with neutral and class cards weighted equally in being selected. By default, the chosen card is then generated and added to the player's hand, unless otherwise specified in the card text.
- Implement 17 cards
    - Making Mummies (ULD_431)
    - Toxic Reinforcements (DRG_255)
    - Messenger Raven (DAL_163)
    - Blazing Battlemage (DRG_239)
    - Hippogryph (DRG_065)
    - Murmy (ULD_723)
    - Candletaker (ULD_205)
    - Bone Wraith (ULD_275)
    - Wasteland Assassin (ULD_274)
    - Colossus of the Moon (ULD_721)
    - Ancestral Guardian (ULD_207)
    - Restless Mummy (ULD_206)
    - Arcane Servant (DAL_092)
    - Mana Reservoir (DAL_748)
    - Hench-Clan Sneak (DAL_090)
    - Dalaran Crusader (DAL_085)
    - Walking Fountain (DAL_047)
- Create task classes
    - `QuestProgressTask`: This class represents the task for processing quest progress.
    - `DiscoverTask`: This class represents the task for running 'Discover' effect.
- Add several methods
    - `HasReborn()`, `SummonReborn()`, `IsQuest()`, `IncreaseQuestProgress()`, `GetQuestProgress()`, `GetQuestProgressTotal()`, `IsSidequest()`, `GetStandardCards()`, `GetWildCards()`, `CreateChoiceCards()`, `GetChoiceCards()`, `HasLifesteal()`

### CODE REVIEW

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  13
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  19
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  19
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

File: Sources/Rosetta/Loaders/QuestLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+int QuestLoader::GetQuestProgressTotal(const std::string_view& cardID)
+{
+    if (cardID == "ULD_431")
+    {
+        return 5;
+    }
+    else if (cardID == "DRG_255")
+    {
+        return 3;
+    }
```
[the remark this answers is not in the document]
Comment (utilForever) [reply Sources/Rosetta/Loaders/QuestLoader.cpp]: if-else는 제어문이고 std::map은 자료형인데 무슨 소리인가요...?

[the remark this answers is not in the document]
Comment (utilForever) [reply Sources/Rosetta/Loaders/QuestLoader.cpp]: 대화로 논의했고 퀘스트가 좀 더 추가되면 고민해 보기로 했습니다.

File: Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp [hunk 2]
```diff
@@ -6,7 +6,94 @@
...
+
+    Player* curPlayer = game.GetCurrentPlayer();
+    Player* opPlayer = game.GetOpponentPlayer();
+    curPlayer->SetTotalMana(10);
+    curPlayer->SetUsedMana(0);
+    opPlayer->SetTotalMana(10);
+    opPlayer->SetUsedMana(0);
+
+    auto& curField = *(curPlayer->GetFieldZone());
+    // const auto curSecret = curPlayer->GetSecretZone();
```
[the remark this answers is not in the document]
Comment (utilForever) [reply Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp]: 수정했습니다.

File: Sources/Rosetta/Tasks/SimpleTasks/DiscoverTask.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+                if (std::find(result.begin(), result.end(), pick) ==
+                    result.end())
+                {
+                    break;
+                }
+            }
+
+            result.emplace_back(pick);
+        }
```
[the remark this answers is not in the document]
Comment (utilForever) [reply Sources/Rosetta/Tasks/SimpleTasks/DiscoverTask.cpp]: 수정했습니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  19
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/UnitTests/CardSets/DalaranCardsGenTests.cpp  17
- Tests/UnitTests/CardSets/UldumCardsGenTests.cpp  19
- Sources/Rosetta/CardSets/DragonsCardsGen.cpp  1
- Tests/UnitTests/CardSets/DragonsCardsGenTests.cpp  31
- Tests/UnitTests/Tasks/PlayerTasks/AttackTaskTests.cpp  8
- Sources/Rosetta/Tasks/SimpleTasks/QuestProgressTask.cpp  3
- Sources/Rosetta/CardSets/UldumCardsGen.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=4932170&bid=16312553)

File: Sources/Rosetta/Loaders/QuestLoader.cpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    return 0;
```
Comment (FYLSunghwan) [inline Sources/Rosetta/Loaders/QuestLoader.cpp]: 현재 퀘스트 진행상황을 GameTag로 저장하는 것처럼 하지 않고 퀘스트 총 진행상황 갯수만 따로 분리한 이유가 있나요?
카드 구현할 때 한번에 다룰 수 있으면 좋겠어요 :D

File: Sources/Rosetta/Tasks/SimpleTasks/SummonTask.cpp [hunk 5]
```diff
@@ -109,6 +109,20 @@ int SummonTask::GetPosition(Entity* source, SummonSide side)
...
+        {
+            const auto tgt = dynamic_cast<Playable*>(target);
+            if (tgt != nullptr)
+            {
+                summonPos = tgt->GetZonePosition() + 1;
+            }
+            else
+            {
+                throw std::invalid_argument(
+                    "SummonTask::Impl() - target is nullptr");
```
Comment (FYLSunghwan) [inline Sources/Rosetta/Tasks/SimpleTasks/SummonTask.cpp]: 얘만 target의 T가 소문자인 것 같습니다!

File: Sources/Rosetta/Loaders/QuestLoader.cpp (hunk 4 above)
Comment (utilForever) [reply Sources/Rosetta/Loaders/QuestLoader.cpp]: 해당 방식을 고려하지 않은 것은 아닙니다만, 지금까지 나온 퀘스트 카드 수가 그리 많지 않아서 따로 추가하는 방식으로 구현했습니다. 충분히 논의할 수 있는 부분이긴 하네요.

File: Sources/Rosetta/Tasks/SimpleTasks/SummonTask.cpp (hunk 5 above)
Comment (utilForever) [reply Sources/Rosetta/Tasks/SimpleTasks/SummonTask.cpp]: 이건 위에 있는 `const auto tgt = dynamic_cast<Playable*>(target);`이 nullptr임을 표시하는 거라서 잘못된 부분이 아닙니다.

File: Sources/Rosetta/Loaders/QuestLoader.cpp (hunk 4 above)
Comment (utilForever) [reply Sources/Rosetta/Loaders/QuestLoader.cpp]: 음, 아니면 playReqs/entourages/questProgressTotal 정보를 하나의 구조체로 정의하고 이걸 넘기는 형태로 작업하는 것도 괜찮을 거 같습니다. 이 작업은 이슈로 등록한 뒤 리팩토링 작업으로 하면 좋겠습니다.

**Pull request merged**